automerge/src/query/list_range.rs

```diff
@@ -0,0 +1,55 @@
+use crate::op_tree::OpTreeNode;
+use crate::types::{Key, OpId};
+use crate::Value;
+use std::fmt::Debug;
+use std::ops::RangeBounds;
+
+#[derive(Debug)]
+pub(crate) struct ListRange<'a, R: RangeBounds<usize>> {
+    range: R,
+    index: usize,
+    pos: usize,
+    last_key: Option<Key>,
+    next_result: Option<(usize, Value<'a>, OpId)>,
+    index_back: usize,
+    root_child: &'a OpTreeNode,
+}
+
+impl<'a, R: RangeBounds<usize>> ListRange<'a, R> {
+    pub(crate) fn new(range: R, root_child: &'a OpTreeNode) -> Self {
+        Self {
+            range,
+            index: 0, // FIXME root_child.seek_to_pos(range.start)
+            pos: 0,   // FIXME range.start
+            last_key: None,
+            next_result: None,
+            index_back: root_child.len(),
+            root_child,
+        }
+    }
+}
+
+impl<'a, R: RangeBounds<usize>> Iterator for ListRange<'a, R> {
+    type Item = (usize, Value<'a>, OpId);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut result = None;
+        for i in self.index..self.index_back {
+            let op = self.root_child.get(i)?;
+            self.index += 1;
+            if op.visible() {
+                if self.range.contains(&self.pos) {
+                    result = self.next_result.replace((self.pos, op.value(), op.id));
+                }
+                if Some(op.key) != self.last_key {
```

Comment:
Apparently, this condition is at fault. It seems to me like we would have to differentiate between inserts and puts here? 

```suggestion
                if op.insert || Some(op.key) != self.last_key {
```

With this, both the tests you added, and mine, pass.

Reply:
Excellent find! - actually the problem is I based this on the range code which using maps relies on keys - list ops are a little more complicated and we need to go off `op.elemid()`.  So glad you caught this.